Mark security properities experimental

milleruntime opened the pull request:
* Make the instance.security properties boolean instead of a classname
* This will remove security interfaces from the properties without
losing the functionality provided.
* The instance.security.authenticator, instance.security.authorizor
and instance.security.permissionHandler properties used interfaces that
contain internal types and were not friendly for public use.

ctubbsii reviewed (changes requested):
@milleruntime These interfaces were added in ACCUMULO-259 (I still have the JIRA issue memorized, because it was quite an ordeal at the time) to be pluggable. It was not merely added to support Kerberos/SASL authentication as an alternative. Rather, it was intended to be completely pluggable... so you could swap in an LDAP implementation, for example.

I don't think these changes you've made in this PR are aligned with the original intention.

Rather, what I think we need is a complete overhaul of these interfaces, to bring them up to speed, so they are public API, still pluggable, and more convenient to swap in implementations.

Until that comprehensive work is done, I think these should remain non-public API, and left as an advanced / marginally supported feature through the "internal" APIs. If it's "easy" to make the existing interfaces themselves public API / SPI, without exposing additional internal types, I'd be okay with that change also.

But, I don't think it's adequate to simply limit things to the two available implementations, as users may have implemented their own in their internal forked versions of Accumulo.

milleruntime commented:
> Until that comprehensive work is done, I think these should remain non-public API, and left as an advanced / marginally supported feature through the "internal" APIs. If it's "easy" to make the existing interfaces themselves public API / SPI, without exposing additional internal types, I'd be okay with that change also.

I made attempts at trying to modify them to fit our API or SPI.  The SPI may be more feasible but the main problem is their implementation relies heavily on access to internal ZooKeeper objects.  Most of the work the classes are doing is just interacting with zookeeper.  The Kerberos impl still uses the ZK impl but with added checks.  So I don't think there is an "easy" approach.

milleruntime commented:
> @milleruntime These interfaces were added in ACCUMULO-259 (I still have the JIRA issue memorized, because it was quite an ordeal at the time) to be pluggable. It was not merely added to support Kerberos/SASL authentication as an alternative. Rather, it was intended to be completely pluggable... so you could swap in an LDAP implementation, for example.

Do you remember if there was an actual customer need to have pluggable LDAP or something other than Kerberos?  Reading the JIRA tickets, it seems this was just a developer idea.  https://issues.apache.org/jira/browse/ACCUMULO-259.

ctubbsii commented:
> Do you remember if there was an actual customer need to have pluggable LDAP or something other than Kerberos? Reading the JIRA tickets, it seems this was just a developer idea. https://issues.apache.org/jira/browse/ACCUMULO-259.

I have been informed in the past that people were doing this. However, I do not know if they still are. Either way, it is a capability I'd like to preserve, and improve upon, even if the API isn't stable. If we want to mark it as `@Experimental` in the interim, due to its limited support as non-public API, I think that might be our best option.

milleruntime commented:
> If we want to mark it as `@Experimental` in the interim, due to its limited support as non-public API, I think that might be our best option.

I agree.  Changed the title and pushed this change.  I added a comment next to the @Experimental tag to try and explain why the tag was added.
